src/Kiota.Builder/CodeDOM/CodeNamespace.cs

```diff
@@ -29,6 +31,26 @@ public override string Name
             StartBlock.Name = name;
         }
     }
+
+    // 
+    private ConcurrentDictionary<string, byte> GroupedElements { get; set; } = new();
+    private ConcurrentDictionary<string, List<CodeElement>> CodeFile { get; set; } = new();
+
+    public void AddToFile(CodeElement rootElement, CodeElement element)
+    {
+        GroupedElements.TryAdd(element.Name, 0);
+
+        var fileElements = CodeFile.GetOrAdd(rootElement.Name, new List<CodeElement> { rootElement });
+        fileElements.Add(element);
+        CodeFile[rootElement.Name] = fileElements;
+    }
+
+    public bool IsRootElement(CodeElement element) =>
+        !GroupedElements.ContainsKey(element.Name);
+
+    public CodeElement[] GetFileElements(CodeElement element) =>
+        CodeFile.ContainsKey(element.Name) ? CodeFile[element.Name].ToArray() : new[] { element };
+
```

Comment:
Instead of introducing this other "view" of the code dom which might get out of sync with the "main" tree by accident when things are moved around, I'd be in favor or introducing a CodeFile element (inherits from code block).
Besides maintaining a single tree, the reasoning is that grouping elements into a single file impacts import paths for some languages (TypeScript, Ruby, PHP, Python...) and aspects needs to be considered by the writer without having to do "complex" calculations on a separate list etc.
Imports of a file will also need to be deduplicated.

Reply:
Done